### PR TITLE
Add PSR12 ruleset

### PR DESCRIPTION
**Pull request by julienfalque**
Ref: #4502.

### CODE REVIEW

**Comment by drupol** [comment]
Can't wait for this :)

**Comment by drupol** [comment]
Just FYI,

On my side, I'm using [a package of mine](https://github.com/drupol/phpcsfixer-configs-php) implementing this ruleset among some others.

This is completely in sync with [PHPCS](https://github.com/squizlabs/PHP_CodeSniffer) and [ECS](https://packagist.org/packages/symplify/easy-coding-standard).

The scripts i'm using in `composer.json`:
```
    "scripts": {
      "phpcsfixer": "vendor/bin/php-cs-fixer  --dry-run --allow-risky=yes --config=./vendor/drupol/php-conventions/config/psr12/php_cs_fixer.config.php --using-cache=no --verbose fix",
  }
```

**Comment by julienfalque** [comment]
Thanks @drupol, I updated the ruleset :)

**Comment by drupol** [comment]
Great, glad it helped :)

**Comment by drupol** [comment]
Weird issue.

**Comment by julienfalque** [comment]
Our tests prevent having a ruleset that overrides a rule with the exact same config.

**Comment by julienfalque** [comment]
@nicolas-grekas @fabpot This introduces changes in the `@Symfony` ruleset, are you ok with those?

**Comment by nicolas-grekas** [comment]
I ran this PR on the Symfony, there are some suspicious changes, but looks good otherwise.

Here are the changes that I think should not happen:

image

image

image

**Comment by drupol** [comment]
Hi,

Just for the record, here are the rules (and the ruleset that uses them) from your screenshots.

> image

`ordered_imports [@Symfony, @PhpCsFixer]`

> image

`full_opening_tag [@PSR1, @PSR2, @Symfony, @PhpCsFixer]`

> image

`global_namespace_import`

**Comment by julienfalque** [comment]
@nicolas-grekas Having trait imports at the very beginning of classes is required by PSR-12, do you mean you don't want Symfony's coding style to follow that specific rule?

The second change looks like a bug unrelated to this PR, `<?=` is supposed to be allowed. In which file was that change introduced?

The last change looks unrelated to this PR, `global_namespace_import` isn't part of any ruleset.

**Comment by fabpot** [comment]
Having trait imports at the beginning of the classes looks ok to me.

**Comment by kubawerlos** [comment]
@nicolas-grekas @julienfalque last change is related to ["strict" flag](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4679#issuecomment-567024804) for NativeConstantInvocationFixer which will be released in `2.17`.

**Comment by nicolas-grekas** [comment]
> last change is related to "strict" flag for NativeConstantInvocationFixer which will be released in 2.17.

OK, then that's something that `NativeConstantInvocationFixer` should *not* do for some selected constants. `PHP_INT_SIZE` is one of them: while for most other constants we don't care, this one, but also e.g. `PHP_VERSION_ID`, is used by PHP at compile time to optimize the compiled opcodes. Removing the `\` breaks the optimization.

I think looking at which constants have a leading `\` in the codebase could give the list. `PHP_DIRECTORY_SEPARATOR` is one of them.

**Comment by kubawerlos** [comment]
Like these https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.3/src/RuleSet.php#L193?

The only update in 2.17 would be making it strict.

**Comment by nicolas-grekas** [comment]
Yes, this. PHP_INT_SIZE should be added to the list.

File: src/RuleSet.php [hunk 1]
```diff
@@ -530,6 +565,13 @@ private function getSetDefinition($name)
             throw new \InvalidArgumentException(sprintf('Set "%s" does not exist.', $name));
         }
 
+        if (\PHP_VERSION_ID < 70100) {
```
Comment (GrahamCampbell) [inline src/RuleSet.php:568]: I don't think presets should be dependant on the runtime PHP version. It need not be the same as the minimum source target version.

Comment (julienfalque) [reply src/RuleSet.php]: We agree. This needed #4946 which is merged now so I removed the condition.

File: tests/Fixtures/Integration/set/@PSR12_php70.test [hunk 2]
```diff
@@ -0,0 +1,6 @@
+--TEST--
+Integration of @PSR12 [PHP 7.0 version].
+--RULESET--
+{"@PSR12": true}
+--REQUIREMENTS--
+{"php": 70000}
```
Comment (mvorisek) [inline tests/Fixtures/Integration/set/@PSR12_php70.test:6]: will be nice to add EOL

File: tests/Fixtures/Integration/set/@PSR12_php71.test [hunk 3]
```diff
@@ -0,0 +1,6 @@
+--TEST--
+Integration of @PSR12 [PHP 7.1 version].
+--RULESET--
+{"@PSR12": true}
+--REQUIREMENTS--
+{"php": 70100}
```
Comment (mvorisek) [inline tests/Fixtures/Integration/set/@PSR12_php71.test:6]: will be nice to add EOL

File: tests/Fixtures/Integration/set/@PSR12_php70.test (hunk 2 above)
Comment (julienfalque) [reply tests/Fixtures/Integration/set/@PSR12_php70.test]: Done, thanks.

File: tests/Fixtures/Integration/set/@PSR12_php71.test (hunk 3 above)
Comment (julienfalque) [reply tests/Fixtures/Integration/set/@PSR12_php71.test]: Done, thanks.

File: tests/Fixtures/Integration/set/@PSR12_php70.test (hunk 2 above)
Comment (julienfalque) [reply tests/Fixtures/Integration/set/@PSR12_php70.test]: Done, thanks.

File: src/RuleSet.php [hunk 4]
```diff
@@ -56,12 +56,53 @@ final class RuleSet implements RuleSetInterface
             'switch_case_space' => true,
             'visibility_required' => true,
         ],
-        '@Symfony' => [
+        '@PSR12' => [
```
Comment (GrahamCampbell) [inline src/RuleSet.php:59]: The `braces` fixer should be configured to allow single_line closure, rather than inheriting the defaults from the PSR2 preset.

Comment (GrahamCampbell) [reply src/RuleSet.php:59]: `braces => ['allow_single_line_closure' => true]`

Comment (GrahamCampbell) [reply src/RuleSet.php:59]: We're also missing `'concat_space' => ['spacing' => 'one']`.

Comment (GrahamCampbell) [reply src/RuleSet.php:59]: `'function_typehint_space' => true` and `'no_leading_namespace_whitespace' => true` are also missing and should be included (I think).

File: src/RuleSet.php [hunk 5]
```diff
@@ -56,12 +56,53 @@ final class RuleSet implements RuleSetInterface
...
             '@PSR2' => true,
-            'array_syntax' => ['syntax' => 'short'],
-            'backtick_to_shell_exec' => true,
             'binary_operator_spaces' => true,
```
Comment (GrahamCampbell) [inline src/RuleSet.php:61]: This is too strict. The spec says at least one space on the right, and not exactly one.

**Comment by GrahamCampbell** [comment]
Left you a PR at https://github.com/julienfalque/PHP-CS-Fixer/pull/10. :)

File: src/RuleSet.php (hunk 5 above)
Comment (GrahamCampbell) [reply src/RuleSet.php:61]: I think we need to add another TODO to do the psr-12 compat issue.

File: src/RuleSet.php [hunk 6]
```diff
@@ -56,12 +56,53 @@ final class RuleSet implements RuleSetInterface
...
+                    'const',
+                ],
+                'sort_algorithm' => 'none',
+            ],
+            'return_type_declaration' => true,
+            'short_scalar_cast' => true,
+            'single_blank_line_before_namespace' => true,
+            'single_trait_insert_per_statement' => true,
+            'ternary_operator_spaces' => true,
+            'unary_operator_spaces' => true,
```
Comment (GrahamCampbell) [inline src/RuleSet.php:89]: Actually, PSR-12 says nothing about the space between an `!` and an expression, so this rule is too strict.

File: src/RuleSet.php (hunk 4 above)
Comment (julienfalque) [reply src/RuleSet.php:59]: > The `braces` fixer should be configured to allow single_line closure

I didn't find anything in the spec that dictates this. Could you please point me at the correct place?

> the spec has an example allowing `{}` body for anonymous classes

Added option `allow_single_line_anonymous_class_with_empty_body`.

> We're also missing `'concat_space' => ['spacing' => 'one']`

Actually adding it would be too strict for [the same reason you mentioned for `binary_operator_spaces`](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4943#discussion_r460592408).

> `'function_typehint_space' => true` and `'no_leading_namespace_whitespace' => true` should be included

I also didn't find anything in the spec for this.

File: src/RuleSet.php (hunk 5 above)
Comment (julienfalque) [reply src/RuleSet.php]: Removed.

File: src/RuleSet.php (hunk 4 above)
Comment (GrahamCampbell) [reply src/RuleSet.php:59]: > I didn't find anything in the spec that dictates this. Could you please point me at the correct place?

Yeh, I can't see it now. Maybe they took it out of the final thing. I remember it was certainly discussed.

> I also didn't find anything in the spec for this.

`function_typehint_space` is covered by section 4.4 and 4.5. They don't spell it out in text, but they say "Note the placement of parentheses, commas, spaces, and braces". I think they did intend to say it, they just didn't write it down properly.

> no_leading_namespace_whitespace

I think this one is implied by their examples, where they are trying to define the spec by examples, instead of by spelling out the rules. Kinda a shame they did that.

Comment (GrahamCampbell) [reply src/RuleSet.php:59]: I've updated my PR again.

Comment (julienfalque) [reply src/RuleSet.php:59]: Rules inferred from examples might be subject to interpretation. I'm not sure we should add those. @SpacePossum @kubawerlos, what do you think?

Comment (GrahamCampbell) [reply src/RuleSet.php:59]: I totally agree the spec is not clear enough too, and we could even go further as to seek an clarification from the authors and/or a spec amendment. It's a shame they didn't get php-cs-fixer and phpcs authors to help write the spec...

Comment (drupol) [reply src/RuleSet.php:59]: I definitely agree, I'm asking myself the same question.

File: tests/Fixtures/Integration/set/@PSR12.test-out.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    use BarTrait;
+
+    public $aaa = 1;
+    public $bbb = 2;
+
+    public function sampleFunction($a, $arg1, $arg2, $arg3, $foo, $b = null)
+    {
+        if ($a === $b) {
+            bar();
+        } elseif ($a    >    $b) {
```
Comment (mvorisek) [inline tests/Fixtures/Integration/set/@PSR12.test-out.php:24]: there should be one space

Comment (GrahamCampbell) [reply tests/Fixtures/Integration/set/@PSR12.test-out.php:24]: ?

Comment (GrahamCampbell) [reply tests/Fixtures/Integration/set/@PSR12.test-out.php:24]: Oh, that's an outstanding issue out of scope of this PR. We already discussed this today. :)

Comment (GrahamCampbell) [reply tests/Fixtures/Integration/set/@PSR12.test-out.php:24]: PSR-12 only insists on exactly one space to the left, but at least one on the right. php-cs-fixer can't do that at the moment.

File: tests/RuleSetTest.php [hunk 8]
```diff
@@ -772,7 +776,7 @@ private static function assertSameRules(array $expected, array $actual, $message
      *
      * Sometimes keys are all string, sometimes they are integers - we need to account for that.
      */
-    private function sort(array &$data)
+    private function sort(array &$data, array $parentPathParts = [])
```
Comment (keradus) [inline tests/RuleSetTest.php:779]: is this change still needed?

**Comment by GrahamCampbell** [comment]
I have actually already re-built this PR myself using the new rule set classes. Would you like me to PR the replacement?

**Comment by keradus** [comment]
If you have sth ready out of the box, please share it and maybe collaborate with Julien to finalize the topic.

What I would love to have is rather agreeable set that we can easily merge and use rather than starting another never ending discussion what config should be inside. Let's take minimum agreeable config, merge, and have needed discussion as 2nd step

**Comment by GrahamCampbell** [comment]
I don't have time to send a replacement PR this second, but in the meanwhile, here's the patch: https://gist.github.com/GrahamCampbell/4902e84e782d2aee1954a25c42aa61ee (without the docs directory changes, and also without correct co-authorship).

**Comment by GrahamCampbell** [comment]
If nobody else gets to this before me (probably in the next 3 weeks), then I will go the whole way and send a PR with docs dir genned, etc.

File: tests/RuleSetTest.php (hunk 8 above)
Comment (julienfalque) [reply tests/RuleSetTest.php]: I forgot about it while fixing merge conflicts... and basically rewrote a similar thing to make tests pass, so yes :)

**Comment by drupol** [comment]
Can't believe this is happening! Great!

**Comment by GrahamCampbell** [comment]
@drupol Just because this ruleset is named PSR12, it doesn't mean it actually enforces all the rules. In fact, many fixers remain to be implemented/updated.

**Comment by GrahamCampbell** [comment]
I think the import fixing will be something many people will notice.

**Comment by drupol** [comment]
> @drupol Just because this ruleset is named PSR12, it doesn't mean it actually enforces all the rules. In fact, many fixers remain to be implemented/updated.

Yes, I guess.

That said, it's nice that we will have the `PSR12` ruleset! It will be refined and improved with time.

**Comment by pkruithof** [comment]
> Just because this ruleset is named PSR12, it doesn't mean it actually enforces all the rules.

Out of curiosity: why not? I would assume it does, otherwise why give it that name?

**Comment by GrahamCampbell** [comment]
> Out of curiosity: why not?

> many fixers remain to be implemented/updated.

**Comment by pkruithof** [comment]
Thanks for the clarification, I understand now and I appreciate all the work you guys are doing here.

File: src/RuleSet/Sets/PSR12RiskySet.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        return [
+            'no_trailing_whitespace_in_string' => true,
+            'no_unreachable_default_argument_value' => true,
+        ];
+    }
+
+    public function getDescription()
+    {
+        return 'Rules that follow `PSR-2 <https://www.php-fig.org/psr/psr-2/>`_ standard.';
```
Comment (keradus) [inline src/RuleSet/Sets/PSR12RiskySet.php:32]:
```suggestion
        return 'Rules that follow `PSR-12 <https://www.php-fig.org/psr/psr-12/>`_ standard.';
```

**Comment by keradus** [comment]
Thank you @julienfalque.

File: tests/Fixtures/Integration/set/@PSR12.test-out.php (hunk 7 above)
Comment (nevmerzhitsky) [reply tests/Fixtures/Integration/set/@PSR12.test-out.php:24]: > ... exactly one space to the left, but at least one on the right.

Seems the same logic used for the left and the right parts: https://www.php-fig.org/psr/psr-12/#62-binary-operators
